### PR TITLE
Add secure node tests

### LINKED ISSUE #7935: Secure websocket proxying
There should be no need to use an insecure protocol for websocket proxying.

### PR DESCRIPTION
**Pull request by raju249**
**Thanks for contributing to Selenium!**
**A PR well described will help maintainers to quickly review and merge it**

Before submitting your PR, please check our [contributing](https://github.com/SeleniumHQ/selenium/blob/trunk/CONTRIBUTING.md) guidelines.
Avoid large PRs, help reviewers by making them as simple and short as possible.

### Description
Fixes https://github.com/SeleniumHQ/selenium/issues/7935

### Types of changes

- [-] Bug fix (non-breaking change which fixes an issue)
- [-] New feature (non-breaking change which adds functionality)
- [-] Breaking change (fix or feature that would cause existing functionality to change)

### Checklist

- [x] I have read the [contributing](https://github.com/SeleniumHQ/selenium/blob/trunk/CONTRIBUTING.md) document.
- [-] My change requires a change to the documentation.
- [-] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: java/server/test/org/openqa/selenium/grid/router/ProxyCdpTest.java [hunk 1]
```diff
@@ -126,6 +128,41 @@ public void onText(CharSequence data) {
     socket.close();
   }
 
+  @Test
+  public void shouldBeAbleToSendMessagesOverSecureWebSocket()
+      throws URISyntaxException, InterruptedException {
+    HttpClient.Factory clientFactory = HttpClient.Factory.createDefault();
+
+    Config secureConfig = new MapConfig(ImmutableMap.of("node", ImmutableMap.of("--self-signed-https","","https-private-key", "", "https-certificate", "")));
```
Comment (shs96c) [inline java/server/test/org/openqa/selenium/grid/router/ProxyCdpTest.java]: This isn't how to set the self-signed cert flag. You need:

```java
ImmutableMap.of(
  "server", ImmutableMap.of(
    "https-self-signed", true));
```

TL;DR: the flags are different from the Config settings. You can see the mapping in `BaseServerFlags`

File: java/server/test/org/openqa/selenium/grid/router/ProxyCdpTest.java [hunk 2]
```diff
@@ -126,6 +128,41 @@ public void onText(CharSequence data) {
...
+
+    ProxyCdpIntoGrid proxy = new ProxyCdpIntoGrid(clientFactory, sessions);
+    proxyServer = new NettyServer(new BaseServerOptions(secureConfig), nullHandler, proxy).start();
+
+    Server<?> backend = createBackendServer(new CountDownLatch(1), new AtomicReference<>(), "Asiago");
```
Comment (shs96c) [inline java/server/test/org/openqa/selenium/grid/router/ProxyCdpTest.java]: Comment: feel free to pick a cheese of your own choice :)

File: java/server/test/org/openqa/selenium/grid/router/ProxyCdpTest.java [hunk 3]
```diff
@@ -126,6 +128,41 @@ public void onText(CharSequence data) {
...
+            latch.countDown();
+          }
+        });
+
+    socket.sendText("Cheese!");
+
+    assertThat(latch.await(5, SECONDS)).isTrue();
+    assertThat(text.get()).isEqualTo("Asiago");
+
+    socket.close();
```
Comment (shs96c) [inline java/server/test/org/openqa/selenium/grid/router/ProxyCdpTest.java]: Use the `socket` in a try-with-resources to ensure it actually closes.

**Review by shs96c — DISMISSED** [review]

File: java/server/test/org/openqa/selenium/grid/router/ProxyCdpTest.java [hunk 4]
```diff
@@ -31,12 +31,7 @@
 import org.openqa.selenium.grid.sessionmap.local.LocalSessionMap;
 import org.openqa.selenium.netty.server.NettyServer;
 import org.openqa.selenium.remote.SessionId;
-import org.openqa.selenium.remote.http.HttpClient;
-import org.openqa.selenium.remote.http.HttpHandler;
-import org.openqa.selenium.remote.http.HttpRequest;
-import org.openqa.selenium.remote.http.HttpResponse;
-import org.openqa.selenium.remote.http.TextMessage;
-import org.openqa.selenium.remote.http.WebSocket;
+import org.openqa.selenium.remote.http.*;
```
Comment (diemol) [inline java/server/test/org/openqa/selenium/grid/router/ProxyCdpTest.java]: Could you please leave the explicit imports?

File: java/server/test/org/openqa/selenium/grid/router/ProxyCdpTest.java [hunk 5]
```diff
@@ -126,9 +132,40 @@ public void onText(CharSequence data) {
...
+        public void onText(CharSequence data) {
+          text.set(data.toString());
+          latch.countDown();
+        }
+      })) {
+      socket.sendText("Cheese!");
+
+      assertThat(latch.await(5, SECONDS)).isTrue();
+      assertThat(text.get()).isEqualTo("Cheedar");
+    } catch (Exception e) {
```
Comment (diemol) [inline java/server/test/org/openqa/selenium/grid/router/ProxyCdpTest.java]: This `catch` can be removed if we are going to throw it again. (It will be thrown anyway if the `catch` is removed).

File: java/server/test/org/openqa/selenium/grid/router/ProxyCdpTest.java [hunk 6]
```diff
@@ -126,9 +132,40 @@ public void onText(CharSequence data) {
     socket.close();
   }
 
-  private Server<?> createBackendServer(CountDownLatch latch, AtomicReference<String> incomingRef, String response) {
+  @Test
+  public void shouldBeAbleToSendMessagesOverSecureWebSocket()
```
Comment (diemol) [inline java/server/test/org/openqa/selenium/grid/router/ProxyCdpTest.java:137]: Is the test supposed to be failing? I executed it locally and it did not pass. Maybe I am missing some conversation that happened in Slack?

**Review by diemol — CHANGES_REQUESTED** [review]

File: java/server/test/org/openqa/selenium/grid/router/ProxyCdpTest.java (hunk 6 above)
Comment (raju249) [reply java/server/test/org/openqa/selenium/grid/router/ProxyCdpTest.java:137]: Yes, it is not passing, the aim of the issue mentioned in this PR is to make it work. I am working on it @diemol 

Yes, Simon and I had some discussion around it. I will be back Monday after a short staycation 😄 

I will update you on the slack thread then.

Thanks

File: java/client/src/org/openqa/selenium/remote/http/netty/NettyWebSocket.java [hunk 7]
```diff
@@ -56,7 +56,13 @@ private NettyWebSocket(AsyncHttpClient client, org.asynchttpclient.Request reque
 
     try {
       URL origUrl = new URL(request.getUrl());
-      URI wsUri = new URI("ws", null, origUrl.getHost(), origUrl.getPort(), origUrl.getPath(), null, null);
+      String wsScheme = "ws";
+
+      if (origUrl.getProtocol().equals("https")) {
+        wsScheme = "wss";
+      }
```
Comment (diemol) [inline java/client/src/org/openqa/selenium/remote/http/netty/NettyWebSocket.java]:
```suggestion
      String wsScheme = "https".equalsIgnoreCase(origUrl.getProtocol()) ? "wss" : "ws";
```

**Review by diemol — CHANGES_REQUESTED** [review]

**Review by diemol — APPROVED** [review]
Thank you @raju249!
